lib/src/value/calculations.ts

```diff
@@ -0,0 +1,157 @@
+import {hash, List, ValueObject} from 'immutable';
+
+import {Value} from './index';
+import {SassNumber} from './number';
+import {SassString} from './string';
+
+export type CalculationValue =
+  | SassNumber
+  | SassCalculation
+  | SassString
+  | CalculationOperation
+  | CalculationInterpolation;
+
+type CalculationValueIterable = CalculationValue[] | List<CalculationValue>;
+
+function checkUnquotedString(value: CalculationValue): void {
+  if (value instanceof SassString && value.hasQuotes) {
+    throw new Error(`Expected ${value} to be an unquoted string.`);
+  }
+}
+
+export class SassCalculation extends Value {
+  readonly name: string;
+  readonly arguments: List<CalculationValue>;
+
+  private constructor(name: string, args: CalculationValueIterable) {
+    super();
+    this.name = name;
+    this.arguments = List(args);
+  }
+
+  static calc(argument: CalculationValue): SassCalculation {
+    checkUnquotedString(argument);
+    return new SassCalculation('calc', [argument]);
+  }
+
+  static min(args: CalculationValueIterable): SassCalculation {
+    args.forEach(checkUnquotedString);
+    return new SassCalculation('min', args);
+  }
+
+  static max(args: CalculationValueIterable): SassCalculation {
+    args.forEach(checkUnquotedString);
+    return new SassCalculation('max', args);
+  }
+
+  static clamp(
+    min: CalculationValue,
+    value?: CalculationValue,
+    max?: CalculationValue
+  ): SassCalculation {
+    if (value === undefined && max === undefined) {
+      let minString: string;
+      if (min instanceof SassString) {
+        minString = min.text;
+      } else if (min instanceof CalculationInterpolation) {
+        minString = min.value;
+      } else {
+        throw new Error(
+          '`value` and `max` are both undefined, but `min` is not a SassString or CalculationInterpolation.'
+        );
+      }
+      const values = minString.split(',').map(s => {
+        const parsed = parseFloat(s.trim());
+        return isNaN(parsed) ? undefined : new SassNumber(parsed);
+      });
+      const error = new Error(
+        `Expected \`min\` to be a comma-separated list of numbers, got \`${min}\``
+      );
```

Comment:
In retrospect, I think the proposal as written is both unclear and overzealous here. Sass itself allows a `clamp()` with fewer than three arguments as long as any of them is an unquoted string or interpolation, _regardless_ of what it contains. I'll update the spec.